node/node_test.go

```diff
@@ -1133,26 +1205,111 @@ func TestAdminEvents_UnspecifiedAddresses(t *testing.T) {
+				if !ok {
+					return false
+				}
+				return bytes.Equal(completeEv.PostComplete.Smesher, app.signers[0].NodeID().Bytes()) &&
+					bytes.Equal(completeEv.PostComplete.Challenge, shared.ZeroChallenge)
+			},
+			3: func(ev pb.IsEventDetails) bool {
+				startEv, ok := ev.(*pb.Event_PostStart)
+				if !ok {
+					return false
+				}
+				return bytes.Equal(startEv.PostStart.Smesher, app.signers[1].NodeID().Bytes()) &&
+					bytes.Equal(startEv.PostStart.Challenge, shared.ZeroChallenge)
+			},
+			4: func(ev pb.IsEventDetails) bool {
+				completeEv, ok := ev.(*pb.Event_PostComplete)
+				if !ok {
+					return false
+				}
+				return bytes.Equal(completeEv.PostComplete.Smesher, app.signers[1].NodeID().Bytes()) &&
+					bytes.Equal(completeEv.PostComplete.Challenge, shared.ZeroChallenge)
+			},
+			5: func(ev pb.IsEventDetails) bool {
+				// TODO(mafa): this event happens once for each NodeID, but should probably only happen once for all
+				_, ok := ev.(*pb.Event_PoetWaitRound)
+				return ok
+			},
+			6: func(ev pb.IsEventDetails) bool {
+				// TODO(mafa): this event happens once for each NodeID, but should probably only happen once for all
+				_, ok := ev.(*pb.Event_PoetWaitProof)
+				return ok
+			},
+			7: func(ev pb.IsEventDetails) bool {
+				startEv, ok := ev.(*pb.Event_PostStart)
+				if !ok {
+					return false
+				}
+				return bytes.Equal(startEv.PostStart.Smesher, app.signers[0].NodeID().Bytes()) &&
+					!bytes.Equal(startEv.PostStart.Challenge, shared.ZeroChallenge)
+			},
+			8: func(ev pb.IsEventDetails) bool {
+				completeEv, ok := ev.(*pb.Event_PostComplete)
+				if !ok {
+					return false
+				}
+				return bytes.Equal(completeEv.PostComplete.Smesher, app.signers[0].NodeID().Bytes()) &&
+					!bytes.Equal(completeEv.PostComplete.Challenge, shared.ZeroChallenge)
+			},
+			9: func(ev pb.IsEventDetails) bool {
+				startEv, ok := ev.(*pb.Event_PostStart)
+				if !ok {
+					return false
+				}
+				return bytes.Equal(startEv.PostStart.Smesher, app.signers[1].NodeID().Bytes()) &&
+					!bytes.Equal(startEv.PostStart.Challenge, shared.ZeroChallenge)
+			},
+			10: func(ev pb.IsEventDetails) bool {
+				completeEv, ok := ev.(*pb.Event_PostComplete)
+				if !ok {
+					return false
+				}
+				return bytes.Equal(completeEv.PostComplete.Smesher, app.signers[1].NodeID().Bytes()) &&
+					!bytes.Equal(completeEv.PostComplete.Challenge, shared.ZeroChallenge)
+			},
+			11: func(ev pb.IsEventDetails) bool {
+				_, ok := ev.(*pb.Event_AtxPublished)
+				return ok
+			},
+			12: func(ev pb.IsEventDetails) bool {
+				_, ok := ev.(*pb.Event_AtxPublished)
+				return ok
+			},
 		}
-		for idx, ev := range success {
+		for {
 			msg, err := stream.Recv()
 			require.NoError(t, err, "stream %d", i)
-			require.IsType(t, ev, msg.Details, "stream %d, event %d", i, idx)
+			for idx, matcher := range matchers {
+				if matcher(msg.Details) {
+					t.Log("matched event", idx)
+					delete(matchers, idx)
+					break
+				}
+			}
```

Comment:
It should somehow fail if it didn't match any matcher.

Reply:
That would break the test. Additional events in my opinion shouldn't require this test to be updated. There are already additional events that are not asserted for and that I intentionally ignored in the test:

For instance there might be one (or two) `Event_PoetWaitRound` before or between the `Event_AtxPublished` events. If I include them in the test the test only runs for longer and then there might be additional events that are again not accounted for aso.

Reply:
On the other hand, the test will hang if an expected event doesn't happen which will be painful to debug :(

Reply:
Good point, I'll add a fail condition when an expected event is not captured after some time or a certain number of unaccounted for events.

Reply:
Actually there is already a timeout for the stream. All events must be received before the timeout expires (3 Minutes) or the test will fail. So if something is stuck and no events are streamed or one of the expected ones is missing the test will fail with an error after at most 3 minutes with logs that print it failed due to a timeout and which events were already matched.